// sanity-plugin-dashboard-widget-feed/lib/data-adapter.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _client = _interopRequireDefault(require("part:@sanity/base/client"));

var _client2 = _interopRequireDefault(require("@sanity/client"));

var _imageUrl = _interopRequireDefault(require("@sanity/image-url"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = {
  getFeed: (feedQuery, cc) => {
    var query = feedQuery ? feedQuery : "*[ _type == \"feedItem\"]";
    var uri = "/data/query/".concat(cc.dataset ? cc.dataset : 'production', "?query=").concat(encodeURI(query));

    if (cc.projectId) {
      return (0, _client2.default)({
        projectId: cc.projectId,
        dataset: cc.dataset ? cc.dataset : 'production',
        useCdn: cc.useCdn ? cc.useCdn : true
      }).request({
        uri,
        withCredentials: false
      });
    }

    return _client.default.request({
      uri,
      withCredentials: false
    });
  },
  urlBuilder: cc => {
    if (cc.projectId) {
      return (0, _imageUrl.default)((0, _client2.default)({
        projectId: cc.projectId,
        dataset: cc.dataset ? cc.dataset : 'production',
        useCdn: cc.useCdn ? cc.useCdn : true
      }));
    }

    return (0, _imageUrl.default)(_client.default);
  }
};
exports.default = _default;
