// sanity-plugin-dashboard-widget-feed/lib/feed-item.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = FeedItem;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _feedItem = _interopRequireDefault(require("./feed-item.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function FeedItem(_ref) {
  var title = _ref.title,
      imageURL = _ref.imageURL,
      target = _ref.target,
      videoOptions = _ref.videoOptions,
      href = _ref.href,
      byline = _ref.byline,
      description = _ref.description,
      category = _ref.category;
  return _react.default.createElement("a", {
    className: _feedItem.default.root,
    href: href,
    target: target ? target : "_blank",
    rel: "noreferrer noopener"
  }, _react.default.createElement("div", {
    className: _feedItem.default.posterContainer
  }, _react.default.createElement("img", {
    className: _feedItem.default.poster,
    src: imageURL
  }), videoOptions ? videoOptions.noplayer && _react.default.createElement("div", {
    className: _feedItem.default.playIcon
  }, _react.default.createElement("span", null)) : ''), _react.default.createElement("h3", {
    className: _feedItem.default.title
  }, title), _react.default.createElement("div", {
    className: _feedItem.default.textwrapper
  }, _react.default.createElement("div", {
    className: _feedItem.default.byline
  }, byline), _react.default.createElement("div", {
    className: _feedItem.default.description
  }, category && description ? "".concat(category, " - ").concat(description) : category ? {
    category
  } : description ? {
    description
  } : '')));
}

FeedItem.propTypes = {
  title: _propTypes.default.string.isRequired,
  target: _propTypes.default.string,
  videoOptions: _propTypes.default.object,
  imageURL: _propTypes.default.string,
  href: _propTypes.default.string.isRequired,
  hasPlayIcon: _propTypes.default.bool,
  category: _propTypes.default.string,
  byline: _propTypes.default.string.isRequired,
  description: _propTypes.default.string
};
FeedItem.defaultProps = {
  imageURL: null,
  hasPlayIcon: false,
  category: null,
  description: null
};
