// sanity-plugin-dashboard-widget-feed/lib/video-item.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = VideoItem;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _feedItem = _interopRequireDefault(require("./feed-item.css"));

var _reactPlayer = _interopRequireDefault(require("react-player"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function VideoItem(_ref) {
  var title = _ref.title,
      id = _ref.id,
      imageURL = _ref.imageURL,
      href = _ref.href,
      byline = _ref.byline,
      description = _ref.description,
      category = _ref.category,
      videoOptions = _ref.videoOptions;
  var thumbnail = imageURL ? imageURL : true;
  return _react.default.createElement("a", {
    className: _feedItem.default.root
  }, _react.default.createElement("div", {
    className: _feedItem.default.posterContainer
  }, _react.default.createElement(_reactPlayer.default, {
    className: _feedItem.default.reactPlayer,
    url: href,
    width: "100%",
    height: "100%",
    playsinline: true,
    controls: videoOptions.controls,
    pip: videoOptions.pip,
    playing: videoOptions.playing,
    light: videoOptions.light ? thumbnail : ''
  })), _react.default.createElement("h3", {
    className: _feedItem.default.title
  }, title), _react.default.createElement("div", {
    className: _feedItem.default.textwrapper
  }, _react.default.createElement("div", {
    className: _feedItem.default.byline
  }, byline), _react.default.createElement("div", {
    className: _feedItem.default.description
  }, category && description ? "".concat(category, " - ").concat(description) : category ? {
    category
  } : description ? {
    description
  } : '')));
}

VideoItem.propTypes = {
  title: _propTypes.default.string.isRequired,
  posterURL: _propTypes.default.string,
  videoOptions: _propTypes.default.shape({
    controls: _propTypes.default.bool,
    pip: _propTypes.default.bool,
    playing: _propTypes.default.bool,
    light: _propTypes.default.bool
  }),
  href: _propTypes.default.string.isRequired,
  category: _propTypes.default.string,
  byline: _propTypes.default.string.isRequired,
  description: _propTypes.default.string
};
VideoItem.defaultProps = {
  posterURL: null,
  category: null,
  description: null
};
